app/web/src/newhotness/logic_composables/change_set.ts

```diff
@@ -75,24 +114,73 @@ const useApplyChangeSetInner = (ctx: Context) => {
   return { loading, performApply };
 };
 
-const useDisableApplyChangeSetInner = () => {
-  const ctx = inject<Context>("CONTEXT");
-  assertIsDefined(ctx);
+export type ChangeSetApprovalId = string;
 
-  const status = useStatus();
+export interface ChangeSetApprovalRequirement {
+  entityId: Ulid;
+  entityKind: string;
+  requiredCount: number;
+  isSatisfied: boolean;
+  applicableApprovalIds: ChangeSetApprovalId[];
+  approverGroups: Record<string, string[]>;
+  approverIndividuals: string[];
+}
 
-  return computed(
-    () =>
-      ctx.changeSet.value?.status !== ChangeSetStatus.Open ||
-      ctx.onHead.value ||
-      status.value === "syncing",
+export type ApprovalStatus = "Approved" | "Rejected";
+
+export interface ChangeSetApproval {
+  id: ChangeSetApprovalId;
+  userId: UserId;
+  status: ApprovalStatus;
+  isValid: boolean; // is this approval "out of date" based on the checksum
+}
+
+export interface ApprovalData {
+  requirements: ChangeSetApprovalRequirement[];
+  latestApprovals: ChangeSetApproval[];
+}
+
+export const approverForChangeSet = (
+  userId: string,
+  approvalData: ApprovalData,
+) =>
+  approvalData.requirements.some((r) =>
+    Object.values(r.approverGroups)
+      .flat()
+      .concat(r.approverIndividuals)
+      .includes(userId),
   );
+
+export const useApproveOrReject = (ctx: Context) => {
+  const perform = async (status: ApprovalStatus, changeSetId?: ChangeSetId) => {
+    const newCtx = { ...ctx };
+    if (changeSetId) {
+      newCtx.changeSetId = computed(() => changeSetId);
+      newCtx.onHead = computed(() => false); // TODO(nick): handle this in a way that is computed based on the change set id hacked in
+    }
+    const api = useApi(newCtx);
```

Comment:
This is the pattern that @jobelenus and I came up with for "making changes to another change set from this change it". It replaces the anti-pattern